### PR TITLE
Add FakeCrossRef tests for backwards, Fix Layer Norm Backward Decomp

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85471
* #85434
* __->__ #85417

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -1774,8 +1779,7 @@ def test_refs_are_in_decomp_table(self, op):
     "nn.functional.pixel_unshuffle",
 )
 
-fake_striding_skips = (
-    "diag_embed",
+fake_striding_skips = [
```
Comment (mruberry) [inline test/test_ops.py:1782]: Can this add comments for each of these objects describing what they do and how we should maintain them?

File: test/test_ops.py [hunk 2]
```diff
@@ -1887,12 +1905,38 @@ def test_fake_autocast(self, device, dtype, op):
         context = torch.cuda.amp.autocast if device == "cuda" else torch.cpu.amp.autocast
         self._test_fake_helper(device, dtype, op, context)
 
+    @onlyCUDA
+    @ops([op for op in op_db if op.supports_autograd], allowed_dtypes=(torch.float,))
+    @skipOps('TestFakeTensor', 'test_fake_crossref_backward', backward_skips)
```
Comment (mruberry) [inline test/test_ops.py:1910]: Let's add a comment explaining what this test does

File: test/test_ops.py [hunk 3]
```diff
@@ -1796,11 +1800,34 @@ def test_refs_are_in_decomp_table(self, op):
...
+    "linalg.svdvals",
+    "nn.functional.binary_cross_entropy_with_logits",
+    "nn.functional.huber_loss",
+    "nn.functional.logsigmoid",
+    "nn.functional.multilabel_soft_margin_loss",
+    "pca_lowrank",
+    "roll",
+    "svd_lowrank",
+    "sgn",
+]
```
Comment (ezyang) [inline test/test_ops.py:1819]: Are these actually permanent skips?  A lot of these look like they could be xfails

Comment (eellison) [reply test/test_ops.py:1819]: they are xfails actually

File: test/test_ops.py [hunk 4]
```diff
@@ -1887,12 +1907,40 @@ def test_fake_autocast(self, device, dtype, op):
...
+
+            # skip these to speed up tests
+            common_skip_ops = (
+                aten.detach.default,
+                aten.empty_strided.default,
+                aten.copy_.default,
+                aten.is_same_size.default,
+
+            )
+            # TODO: enable check_aliasing, too many failures :/
```
Comment (ezyang) [inline test/test_ops.py:1930]: ALL of the primtorch metas alias incorrectly, they all need to be fixed

File: torch/_subclasses/fake_utils.py [hunk 5]
```diff
@@ -9,17 +10,53 @@
 aten = torch.ops.aten
 
 
+def has_storage(t):
+    try:
+        t.storage()
+        return True
+    except Exception:
+        return False
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:18]: nyeh. This is going to trigger exceptions lol

File: test/test_ops.py (hunk 4 above)
Comment (eellison) [reply test/test_ops.py:1930]: Well, they're not registered to fake tensor right now..

File: torch/_subclasses/fake_utils.py (hunk 5 above)
Comment (eellison) [reply torch/_subclasses/fake_utils.py:18]: i looked into expose `has_storage` as a method on python tensor but i was a little confused about `metaclass_properties`. where should I put it ?

File: torch/_subclasses/fake_utils.py [hunk 6]
```diff
@@ -50,16 +87,41 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
                     )
-                    fake_r = func(*fake_args, **fake_kwargs)
+                    with warnings.catch_warnings():
+                        fake_r = func(*fake_args, **fake_kwargs)
             except UnsupportedFakeTensorException:
                 pass
 
         r = func(*args, **kwargs)
         if fake_r is not None:
+            r_flat = tree_flatten(r)[0]
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:97]: nit: `r_flat, _ = ...`

File: torch/_subclasses/fake_utils.py [hunk 7]
```diff
@@ -50,16 +87,41 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
+            f_flat = tree_flatten(fake_r)[0]
+            assert len(r_flat) == len(
+                r_flat
+            ), f"Mismatch {len(r_flat)} != {len(r_flat)} on {func}"
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:101]: technically you should check the specs too

File: torch/_subclasses/fake_utils.py [hunk 8]
```diff
@@ -50,16 +87,41 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
+
+            if self.check_aliasing:
+                assert outputs_alias_inputs(r, (args, kwargs)) == outputs_alias_inputs(
+                    fake_r, (fake_args, fake_kwargs)
+                ), f"Mismatch on {func}"
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:106]: it's helpful to print what the expected and actual values were

File: torch/_subclasses/fake_utils.py [hunk 9]
```diff
@@ -50,16 +87,41 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
+                    fake_out, torch.Tensor
+                ), f"Mismatched number of tensor outputs on {func}"
+                if r_is_ten:
+                    assert (
+                        r_out.requires_grad == fake_out.requires_grad
+                    ), f"Mismatch on {func}"
+                    if has_storage(r_out):
+                        assert (
+                            r_out.storage_offset() == fake_out.storage_offset()
+                        ), f"Mismatch on {func}"
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:123]: esp for something like this!

File: torch/testing/_internal/composite_compliance.py [hunk 10]
```diff
@@ -399,6 +399,26 @@ def gather_leaf_tensors(args, kwargs):
...
+
+    flat_results, _ = tree_flatten(results)
+    flat_diff_results = [r for r in flat_results if r.requires_grad]
+    assert len(flat_diff_results) > 0
+
+    grads = [torch.ones(r.shape, device=r.device, dtype=r.dtype) for r in flat_diff_results]
+    leaf_tensors = gather_leaf_tensors(args, kwargs)
+    assert len(leaf_tensors) > 0
+    return torch.autograd.grad(flat_diff_results, leaf_tensors,
+                               grads, allow_unused=True, retain_graph=True)
```
Comment (ezyang) [inline torch/testing/_internal/composite_compliance.py:419]: no substantive changes?

**Review by ezyang — APPROVED** [review]

File: test/test_ops.py (hunk 4 above)
Comment (ezyang) [reply test/test_ops.py:1930]: Are you sure? I thought we hit them for dynamic shapes preferentially

File: torch/_subclasses/fake_utils.py (hunk 5 above)
Comment (eellison) [reply torch/_subclasses/fake_utils.py:18]: i guess `THPDtype_properties`..

Comment (ezyang) [reply torch/_subclasses/fake_utils.py:18]: you could bind it as a free standing function in _C, that's how we exposed a lot of "private" tensor internals

File: test/test_ops.py (hunk 4 above)
Comment (eellison) [reply test/test_ops.py:1930]: We do, you're right, but they're disabled for static fake. + 1 motivation for regularizing the path symbolic/fake (not to mention https://github.com/pytorch/pytorch/issues/84308). I have a pr that i had been working on to do this that i havent quite been able to push through finish line.

The aliasing tests are on for the non backward. I'll work on getting them on for backward as a follow up.

File: torch/_subclasses/fake_utils.py (hunk 7 above)
Comment (eellison) [reply torch/_subclasses/fake_utils.py:101]: Filed issue https://github.com/pytorch/pytorch/issues/85438

**Comment by eellison** [comment]
Any idea why this failure didn’t show up on the PR ? Cc @seemethere

**Comment by eellison** [comment]
@clee2000 do you know why these tests are not running ? sort of frustrating

**Comment by clee2000** [comment]
you can add the `ciflow/trunk` label to your pr to kick off trunk tests
